core/trino-spi/src/main/java/io/trino/spi/type/Type.java

```diff
@@ -175,6 +175,38 @@ default Optional<Range> getRange()
         return Optional.empty();
     }
 
+    /**
+     * Returns the maximum value that compares less than {@code value}.
+     * <p>
+     * The type of the value must match {@link #getJavaType}.
```

Comment:
> It can't match getJavaType. For types such as smallint, integer, bigint, getJavaType returns `long.class`. This method takes an Object, so it doesn't support primitive values (they would need to be boxed).

yes, I meant boxing here 

btw this is modeled like we did for getRange in https://github.com/trinodb/trino/commit/e23e317687d5ddede6c1610eb5af35f5e55e5543#diff-3eedf5cda66ac7a68eb48f4eee146a4ccf5658fd07436ba5b279a089294a5c0cR174

> It's only supported for orderable types


yes, same as getRange
non-orderable types will remove `Optional.empty()` since there is no "next" nor "previous"

> The signature is not tight enough to represent all carrier types safely.

this part i did not understand.
@martint  can you please clarify?